rfcs/20190411-tensorboard-improved-plugin-ext.md

```diff
@@ -0,0 +1,151 @@
+import setuptools
+setuptools.setup(
+    name="tensorboard_plugin_widgets",
+    description="Widget chart support for TensorBoard",
+    packages=setuptools.find_packages(),
+    # more metadata (version, license, authors…)
+    entry_points={
+        "tensorboard_plugins": [
+            "widgets = tensorboard_plugin_widgets.plugin:get_plugin",
+        ],
+    },
+)
+```
+
+assuming that the `tensorboard_plugin_widgets.plugin` module exposes a function `get_plugin` conforming to the interface described above.
+
+To facilitate plugin development without requiring plugins to be packaged into a wheel and installed into a virtualenv, TensorBoard will learn an `--extra_plugins` flag, whose value will be a list of comma-separated object references. For instance, for testing the widgets plugin, one might invoke
+
+```
+tensorboard --logdir /tmp/whatever \
+    --extra_plugins tensorboard_plugin_widgets.plugin:get_plugin
+```
+
+with the `tensorboard_plugin_widgets.plugin` module on the Python path.
+
+[entry-points]: https://packaging.python.org/specifications/entry-points/
+[tb-loader]: https://github.com/tensorflow/tensorboard/blob/dc71eeea403eb027612422ab6cd3061d4f9a10f3/tensorboard/plugins/base_plugin.py#L149-L162
+
+### Frontend binary loading
+One of our soft goals is to enable existing visualization suite to integrate with TensorBoard with ease without having to change, for instance, module system. Of course, a plugin has to procure data by writing a backend and some glue code to integrate with our backend, but we are hoping that most of the visualization part can remain untouched.
+
+#### Requirement: Colab Integration
+TensorBoard recently added support for Jupyter and Colab and the feature is well received by the community. The new plugin frontend design should be compatible with Colab integration.
+
+One of the critical limitation Colab imposes TensorBoard is the usage of iframe and it is related to how it routes requests to the backend. To make it transparent to TensorBoard, Colab uses a ServiceWorker to intercept and reroute requests to localhost via its internal API. However, due to the implementation of Colab's ServiceWorker that is designed to intercept only requests to localhost (cannot use a relative path), one cannot render an iframe in an output cell.
+
+This restriction is relevant for *all solutions* below since a plugin can decide to render an iframe that makes a request anywhere.
+
+#### Non-iframe
+Generally speaking, there are largely two types of solutions: one using iframe and one without using an iframe. Within the non-iframe based solution, there are several ways to achieve the goal and they are depicted below.
+
+One critical drawback of all non-iframe approaches is that TensorBoard cannot enforce plugin authors from inadvertently mutating the global object. For instance, if a library that a plugin transitively depends polyfills and monkey-patches a global object, it can influence how other plugins behave.
+
+**Pro**
```

Comment:
See my comment above. The structure of the text around this point is somewhat hard to follow. It's unclear 
1. what the Pro and Con sections refer to
2. why these sections exist for some of the options but not others

Reply:
+1.  What's confusing here is that the options 1, 2, and 3 are at different levels of the hierarchy.  IIUC, the intent is:

- Option A: Non-iFrame
    - Pro, con
    - Option A1
    - Option A2
- Option B: iFrame
    - Pro, con